shared/protocol/src/main/java/io/pravega/shared/segment/SegmentToContainerMapper.java

```diff
@@ -59,6 +61,26 @@ public int getTotalContainerCount() {
      * @return Integer indicating the container id for the given StreamSegment.
      */
     public int getContainerId(String streamSegmentName) {
+        int containerId = Integer.MIN_VALUE;
+        if (!isAddressableSegment(streamSegmentName)) {
```

Comment:
This change in itself will require a perf regression test. This is on the hot path for every operation (including each append) on the segment store.

Let's try to make it as lean as possible.

Reply:
Thanks, that's actually a very good point. I have tested the potential regression locally, and I can see that this PR's logic (i.e., String manipulation and/or hashing) is between 1.5x-3x slower than the existing one (i.e., only hashing). I got that observation from a micro-benchmark I executed locally by running 10 times a 10 million iteration loop in which the `SegmentToContainerMapper` object was executing `getContainerId()` for a metadata and a regular segment on each iteration (so, 20 million operations per round). For this reason, I have added a flag (`enableAdmin`) to the logic of `SegmentToContainerMapper` that only attempts to parse internal metadata segments if the flag is enabled, otherwise it works as always. This is the difference between enabling and not enabling this flag:

```
 Task :shared:protocol:SegmentToContainerMapperTests.main()
Time elapsed (admin mode disabled): 3227
Time elapsed (admin mode enabled): 10154
Time elapsed (admin mode disabled): 4539
Time elapsed (admin mode enabled): 8873
Time elapsed (admin mode disabled): 3842
Time elapsed (admin mode enabled): 9267
Time elapsed (admin mode disabled): 3678
Time elapsed (admin mode enabled): 7369
Time elapsed (admin mode disabled): 4777
Time elapsed (admin mode enabled): 8582
Time elapsed (admin mode disabled): 3267
Time elapsed (admin mode enabled): 9471
Time elapsed (admin mode disabled): 4776
Time elapsed (admin mode enabled): 6697
Time elapsed (admin mode disabled): 5036
Time elapsed (admin mode enabled): 8595
Time elapsed (admin mode disabled): 5763
Time elapsed (admin mode enabled): 9213
Time elapsed (admin mode disabled): 3223
Time elapsed (admin mode enabled): 8996
```
In the distributed mode of the Segment Store service, the `SegmentToContainerMapper` will have this flag set to `true` only if the service is configured with `pravegaservice.admin.gateway.enable=true`. Otherwise, it will work as always. I have also added a note on the `pravegaservice.admin.gateway.enable` flag regarding the potential performance implications.

Reply:
OK, that's a good way to minimize impact. Later on, we should probably see how to optimize this method or even remove it from the hot path.

One idea is to only run it for `createSegment`; otherwise it will just return a `SegmentNotExistsException` if you routed it wrongly - again, an improvement for another time.